### PR TITLE
Component/Online Banking

### PR DESCRIPTION
**Pull request by neelSharma12**
This PR contains implementation for online banking component
OnlineBankingUI
.

### CODE REVIEW

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+    // MARK: - Items
+
+    private var tAndcLink: String {
+        return paymentMethod.type == .onlineBankingCZ ? "https://static.payu.com/sites/terms/files/payu_privacy_policy_cs.pdf" : "https://static.payu.com/sites/terms/files/payu_privacy_policy_sk.pdf"
+    }
+
+    /// The terms and condition message item.
+    // swiftlint:disable line_length
+    internal lazy var tAndcLabelItem: FormAttributedLabelItem = .init(tAndCText: "By clicking continue you agree with the #terms and conditions#",
```
Comment (neelSharma12) [inline AdyenComponents/OnlineBanking/OnlineBankingComponent.swift]: I have to change replace the static string with the localized string

File: Adyen/Core/Payment Methods/Abstract/AnyPaymentMethodDecoder.swift [hunk 2]
```diff
@@ -87,6 +89,11 @@ internal enum AnyPaymentMethodDecoder {
             let brand = try? container.decode(String.self, forKey: .brand)
             let isIssuersList = try container.containsValue(.issuers)
 
+            let paymentDecoder = PaymentMethodType(rawValue: type).map { decoders[$0, default: defaultDecoder] } ?? defaultDecoder
```
Comment (mohammedDehairy) [inline Adyen/Core/Payment Methods/Abstract/AnyPaymentMethodDecoder.swift]: may I ask, why did you move this up here?

File: Adyen/UI/Form/FormItemViewBuilder.swift [hunk 3]
```diff
@@ -40,6 +40,18 @@ public struct FormItemViewBuilder {
         FormRegionPickerItemView(item: item)
     }
     
+    /// Builds `IssuersItemView` from `IssuersItem`.
+    @_spi(AdyenInternal)
+    public func build(with item: FormIssuerListItem) -> FormItemView<FormIssuerListItem> {
+        FormIssuerItemView(item: item)
+    }
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/FormItemViewBuilder.swift]: what's `FormIssuerListItem ` for?

File: Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        self.link = link
+    }
+
+    public var identifier: String?
+
+    /// The style of the label.
+    public var style: TextStyle
+
+    /// The text of the label.
+    public var tAndCText: String
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift]: please avoid abbreviations :-), its hard to read

File: Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+    public func build(with builder: FormItemViewBuilder) -> AnyFormItemView {
+        let label = ADYLabel()
+        label.numberOfLines = 0
+        label.textAlignment = style.textAlignment
+        label.textColor = style.color
+
+        let attributedString = NSMutableAttributedString(string: tAndCText.replacingOccurrences(of: "#", with: " "))
+        let rangeOfTAndCText = getTandCSubstringStartIndexes(mainString: tAndCText)
+        attributedString.addAttribute(.foregroundColor, value: UIColor.Adyen.defaultBlue, range: rangeOfTAndCText)
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift]: please use a separate `TextStyle` for the link :-) and enable merchants to customize it.

File: Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+        label.attributedText = attributedString
+        label.accessibilityIdentifier = identifier
+        label.font = style.font
+        label.backgroundColor = style.backgroundColor
+        label.adyen.round(using: style.cornerRounding)
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift]: please also apply the `non link` `TextStyle` through the attributed string, instead of configuring the UILabel

File: Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(labelTapped))
+        label.addGestureRecognizer(tapGesture)
+        label.isUserInteractionEnabled = true
+        label.addGestureRecognizer(tapGesture)
+        return label
+    }
+
+    @objc private func labelTapped(gesture: UITapGestureRecognizer) {
+        if let url = URL(string: link) {
+            AppLauncher().openCustomSchemeUrl(url, completion: nil)
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift]: please check first if the url is http or custom scheme?

File: Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        }
+    }
+
+    private func getTandCSubstringStartIndexes(mainString: String) -> NSRange {
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift]: please avoid abbreviations :-)

File: Adyen/UI/Form/Items/IssuerList/FormIssuerItemView.swift [hunk 9]
```diff
@@ -0,0 +1,33 @@
+//
+// Copyright (c) 2022 Adyen N.V.
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+import UIKit
+
+internal final class FormIssuerItemView: FormTextItemView<FormIssuerListItem> {
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/Items/IssuerList/FormIssuerItemView.swift]: I believe this class/file is not needed, correct me if I am wrong.

File: Adyen/UI/Form/Items/IssuerList/FormIssuerListItem.swift [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+// Copyright (c) 2022 Adyen N.V.
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+import Foundation
+
+/// Describes a phone number entry form item.
+@_spi(AdyenInternal)
+public final class FormIssuerListItem: FormTextItem {
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/Items/IssuerList/FormIssuerListItem.swift]: I believe this class/file is not needed, correct me if I am wrong.

File: Adyen/UI/Form/Items/Value Pickers/Issuer List Picker/FormIssuerPickerItem.swift [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+import UIKit
+
+@_spi(AdyenInternal)
+public typealias IssuerPickerItem = BasePickerElement<Issuer>
+
+@_spi(AdyenInternal)
+public final class FormIssuersPickerItem: BaseFormPickerItem<Issuer> {
+
+    public override init(preselectedValue: IssuerPickerItem, selectableValues: [IssuerPickerItem], style: FormTextItemStyle) {
+        assert(selectableValues.count > 0)
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/Items/Value Pickers/Issuer List Picker/FormIssuerPickerItem.swift]: please use `AdyenAssertion.assertionFailure(message: "message")`

File: Adyen/UI/Form/Items/Value Pickers/Issuer List Picker/FormIssuerPickerItem.swift [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+import UIKit
+
+@_spi(AdyenInternal)
+public typealias IssuerPickerItem = BasePickerElement<Issuer>
+
+@_spi(AdyenInternal)
+public final class FormIssuersPickerItem: BaseFormPickerItem<Issuer> {
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/Items/Value Pickers/Issuer List Picker/FormIssuerPickerItem.swift:13]: I think this class is redundant, can we just use `BaseFormPickerItem<Issuer> ` directly and make a typealias from `BaseFormPickerItem<Issuer>` to  `FormIssuersPickerItem `

File: Adyen/UI/List/ListItemView.swift [hunk 13]
```diff
@@ -168,7 +167,7 @@ public final class ListItemView: UIView, AnyFormItemView {
 
         trailingTextLabel.setContentHuggingPriority(.required, for: .horizontal)
         
-        NSLayoutConstraint.activate(constraints)
+        NSLayoutConstraint.activate(viewConstraints)
```
Comment (mohammedDehairy) [inline Adyen/UI/List/ListItemView.swift]: if we don't need those changes, can you revert them please?

File: Adyen/UI/View Controllers/SecuredViewController.swift [hunk 14]
```diff
@@ -78,6 +78,10 @@ public final class SecuredViewController: UIViewController {
         delegate?.viewWillAppear(viewController: self)
     }
 
+    public func getChildViewController() -> UIViewController {
```
Comment (mohammedDehairy) [inline Adyen/UI/View Controllers/SecuredViewController.swift]: this can be a var instead of a function, right?

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+    /// Component's configuration
+    public var configuration: Configuration
+
+    private let onlineBankingPaymentMethod: OnlineBankingPaymentMethod
+
+    private var selectedIssuer: Issuer
+
+    // MARK: - Items
+
+    private var tAndcLink: String {
```
Comment (mohammedDehairy) [inline AdyenComponents/OnlineBanking/OnlineBankingComponent.swift]: please avoid abbreviations :-)

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+//
+
+@_spi(AdyenInternal) import Adyen
+import UIKit
+
+/// A component that provides a form for Online Banking payment.
+public final class OnlineBankingComponent: PaymentComponent,
+                                           PresentableComponent,
+                                           ComponentLoader,
+                                           PaymentAware {
```
Comment (mohammedDehairy) [inline AdyenComponents/OnlineBanking/OnlineBankingComponent.swift:14]: I don't think `OnlineBankingComponent ` is a `ComponentLoader `, since it doesn't contain sub components, right?, but it is a `LoadingComponent`

File: AdyenComponents/OnlineBanking/OnlineBankingDetails.swift [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+@_spi(AdyenInternal) import Adyen
+
+/// Contains the details supplied by the Online Banking component.
+public struct OnlineBankingDetails: PaymentMethodDetails {
+
+    /// The payment method type.
+    public let type: PaymentMethodType
+
+    ///  Selected issuer
+    public let issuer: String?
```
Comment (mohammedDehairy) [inline AdyenComponents/OnlineBanking/OnlineBankingDetails.swift]: why is `issuer ` optional?

File: Demo/Configuration.swift [hunk 18]
```diff
@@ -43,7 +43,7 @@ internal enum ConfigurationConstants {
 
     static let demoServerAPIKey = "{YOUR_DEMO_SERVER_API_KEY}"
 
-    static let applePayMerchantIdentifier = "{YOUR_APPLE_PAY_MERCHANT_IDENTIFIER}"
+    static let applePayMerchantIdentifier = "merchant.com.adyen.checkoutsandbox.test"
```
Comment (mohammedDehairy) [inline Demo/Configuration.swift]: please don't forget to revert this change :-)

File: Adyen/UI/Form/FormItemViewBuilder.swift (hunk 3 above)
Comment (neelSharma12) [reply Adyen/UI/Form/FormItemViewBuilder.swift]: This is for issuer list picker.

File: Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift (hunk 4 above)
Comment (neelSharma12) [reply Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift]: Okay.  I will rename this.

File: Adyen/Core/Payment Methods/Abstract/AnyPaymentMethodDecoder.swift (hunk 2 above)
Comment (neelSharma12) [reply Adyen/Core/Payment Methods/Abstract/AnyPaymentMethodDecoder.swift]: sorry, this I was suppose to revert as earlier I added some different code.

File: Adyen/UI/View Controllers/SecuredViewController.swift [hunk 19]
```diff
@@ -78,7 +78,7 @@ public final class SecuredViewController: UIViewController {
         delegate?.viewWillAppear(viewController: self)
     }
 
-    public func getChildViewController() -> UIViewController {
+    public var getChildViewController: UIViewController {
```
Comment (mohammedDehairy) [inline Adyen/UI/View Controllers/SecuredViewController.swift]: why not just expose the childViewController and make it internal or something? :-) instead of this var, since getXX is not a good Swift var name

File: Adyen/UI/Form/Items/IssuerList/FormIssuerItemView.swift (hunk 9 above)
Comment (neelSharma12) [reply Adyen/UI/Form/Items/IssuerList/FormIssuerItemView.swift]: Yes, I have removed them. Thanks for pointing it out.

File: Adyen/UI/Form/Items/IssuerList/FormIssuerListItem.swift (hunk 10 above)
Comment (neelSharma12) [reply Adyen/UI/Form/Items/IssuerList/FormIssuerListItem.swift]: Yes, I have removed them. Thanks for pointing it out :)

File: Adyen/UI/Form/Items/Value Pickers/Issuer List Picker/FormIssuerPickerItem.swift (hunk 12 above)
Comment (neelSharma12) [reply Adyen/UI/Form/Items/Value Pickers/Issuer List Picker/FormIssuerPickerItem.swift:13]: BaseFormPickerItem is an Abstract class that needs to be subclassed.

File: Adyen/UI/List/ListItemView.swift (hunk 13 above)
Comment (neelSharma12) [reply Adyen/UI/List/ListItemView.swift]: yes.

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift (hunk 16 above)
Comment (neelSharma12) [reply AdyenComponents/OnlineBanking/OnlineBankingComponent.swift:14]: yes, you are right :)

File: Adyen/Core/Payment Methods/Abstract/AnyPaymentMethodDecoder.swift [hunk 20]
```diff
@@ -87,6 +89,9 @@ internal enum AnyPaymentMethodDecoder {
             let brand = try? container.decode(String.self, forKey: .brand)
             let isIssuersList = try container.containsValue(.issuers)
 
+            if type == "onlineBanking_CZ" || type == "onlineBanking_SK" {
```
Comment (erenbesel) [inline Adyen/Core/Payment Methods/Abstract/AnyPaymentMethodDecoder.swift]: Could you explain we have to check for type manually here?

File: Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+// Copyright (c) 2022 Adyen N.V.
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+import Foundation
+
+/// Simple form item that represent a single attributed UILabel element.
+@_spi(AdyenInternal)
+public class FormAttributedLabelItem: FormItem {
```
Comment (erenbesel) [inline Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift:12]: The name and description of this class is generic which is nice, but then it has `terms and conditions`property which makes it non generic.
I would assume `termsAndConditionsText` can be renamed to just something like `attributedText`?

File: Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+        return label
+    }
+
+    @objc private func labelTapped(gesture: UITapGestureRecognizer) {
+        if let url = URL(string: link) {
+            UIApplication.shared.open(url, options: [:], completionHandler: nil)
+        }
+    }
+
+    private func getTermsandConditionsSubstringRange(mainString: String) -> NSRange {
```
Comment (erenbesel) [inline Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift]: Let's avoid using the `get` keyword. It's not Apple style :)

(If you remove the `get` part the function's meaning stays the exact same)

File: Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+        let termsAndConditionsText = mainString.split(separator: "#")[1]
+        if let range = mainString.range(of: termsAndConditionsText) {
+            return NSRange(range, in: mainString)
+        }
+        return NSRange(location: 0, length: 0)
+    }
+
+    private func getLinkRangesInString() -> [NSRange] {
```
Comment (erenbesel) [inline Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift]: Same as here :)

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+    private var selectedIssuer: Issuer
+
+    // MARK: - Items
+
+    private var termsAndConditionsLink: String {
+        return paymentMethod.type == .onlineBankingCZ ? "https://static.payu.com/sites/terms/files/payu_privacy_policy_cs.pdf" : "https://static.payu.com/sites/terms/files/payu_privacy_policy_sk.pdf"
+    }
+
+    /// The terms and condition message item.
+    // swiftlint:disable line_length
```
Comment (erenbesel) [inline AdyenComponents/OnlineBanking/OnlineBankingComponent.swift]: If you break lines after each parameter like after `localizationParamters` too, this warning may go away (also if you change the `termAndConditionText` to generic name as I mentioned above, that would help as well).

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+    internal lazy var termsAndConditionsLabelItem: FormAttributedLabelItem = .init(termsAndConditionsText: localizedString(.onlineBankingTermsAndConditions, configuration.localizationParameters),
+                                                                      link: termsAndConditionsLink,
+                                                                      style: configuration.style.footnoteLabel,
+                                                                      linkTextStyle: configuration.style.linkTextLabel,
+                                                                      identifier: ViewIdentifierBuilder.build(scopeInstance: "AdyenDropIn.OnlineBankingComponent", postfix: "OnlineBankingTAndCLabel"))
```
Comment (erenbesel) [inline AdyenComponents/OnlineBanking/OnlineBankingComponent.swift]: On the new components we use a constant enum to contain these view identifiers, you can check`ACHDirectDebitComponent`. 

And `AdyenDropIn` prefix is for DropIn-only components. This looks it can be a stand alone component so we can remove that prefix.

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+        item.title = localizedString(.continueTitle, configuration.localizationParameters)
+        item.buttonSelectionHandler = { [weak self] in
+            self?.didSelectContinueButton()
+        }
+        return item
+    }()
+
+    /// The Issuer List item.
+    internal lazy var issuerListItem: FormIssuersPickerItem = {
+        let defaultIssuer = onlineBankingPaymentMethod.issuers[0]
```
Comment (erenbesel) [inline AdyenComponents/OnlineBanking/OnlineBankingComponent.swift]: Better to be cautious of bounds when accessing arrays. You can use `safeIndex` subscript

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+
+        let issuerListPickerItem: [IssuerPickerItem] = onlineBankingPaymentMethod.issuers.map({
```
Comment (erenbesel) [inline AdyenComponents/OnlineBanking/OnlineBankingComponent.swift]: The name is singular but the result is an array :)

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift [hunk 28]
```diff
@@ -0,0 +1,132 @@
...
+            return IssuerPickerItem(identifier: $0.identifier, element: $0)
+        })
+        let issuerPickerItem = FormIssuersPickerItem(preselectedValue: issuerListPickerItem[0],
```
Comment (erenbesel) [inline AdyenComponents/OnlineBanking/OnlineBankingComponent.swift]: Another direct access without checking size.

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift [hunk 29]
```diff
@@ -0,0 +1,132 @@
...
+                                        selectableValues: issuerListPickerItem,
+                                        style: configuration.style.textField)
+        issuerPickerItem.title = localizedString(.selectFieldTitle, configuration.localizationParameters)
+        issuerPickerItem.identifier = ViewIdentifierBuilder.build(scopeInstance: "AdyenDropIn.OnlineBankingComponent", postfix: "issuersList")
+        _ = issuerPickerItem.publisher.addEventHandler({ issuerPickerITem in
```
Comment (erenbesel) [inline AdyenComponents/OnlineBanking/OnlineBankingComponent.swift]: Would the `observe` function achieve the same thing here? `addEventHandler ` function is one step lower and is not used directly usually. You can check the usages of `observe`

File: Adyen/Core/Payment Methods/Abstract/AnyPaymentMethodDecoder.swift (hunk 20 above)
Comment (neelSharma12) [reply Adyen/Core/Payment Methods/Abstract/AnyPaymentMethodDecoder.swift]: OnlineBanking's payment method response contains issuers list and by default we use IssuerListPaymentMethodDecoder() for the payment methods that contains issuers list. So, I had to apply this check because I wanted to use OnlineBankingPaymentMethodDecoder() method.

File: Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift (hunk 21 above)
Comment (neelSharma12) [reply Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift:12]: Good point, Eren but instead of renaming it to **attributedText** I am renaming to **originalText** because we create the attributed string later.

File: Adyen/Core/Payment Methods/Abstract/AnyPaymentMethodDecoder.swift (hunk 20 above)
Comment (descorp) [reply Adyen/Core/Payment Methods/Abstract/AnyPaymentMethodDecoder.swift]: Maybe move it inside of `if isIssuersList { ... }` ?
Then it would be clear that `onlineBanking_CZ` is a special case of "issuers list" ?

Comment (neelSharma12) [reply Adyen/Core/Payment Methods/Abstract/AnyPaymentMethodDecoder.swift]: Very good point. I have moved it inside the  isIssuersList case

File: Adyen/Core/Payment Methods/Abstract/AnyPaymentMethodDecoder.swift [hunk 30]
```diff
@@ -273,3 +277,9 @@ private struct AtomePaymentMethodDecoder: PaymentMethodDecoder {
         .atome(try AtomePaymentMethod(from: decoder))
     }
 }
+
+private struct OnlineBankingPaymentMethodDecoder: PaymentMethodDecoder {
```
Comment (descorp) [inline Adyen/Core/Payment Methods/Abstract/AnyPaymentMethodDecoder.swift:281]: We keep all decoders in `AnyPaymentMethodDecoder.swift`
Does it need to be here?

File: Adyen/Core/Payment Methods/OnlineBankingPaymentMethod.swift [hunk 31]
```diff
@@ -0,0 +1,65 @@
+//
+// Copyright (c) 2022 Adyen N.V.
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+import Foundation
+
+/// An issuer (typically a bank) in an issuer list payment method.
+public struct Issuer: Decodable, CustomStringConvertible, Equatable {
```
Comment (descorp) [inline Adyen/Core/Payment Methods/OnlineBankingPaymentMethod.swift:10]: Is this the same issuer we have in `IssuerListPaymentMethod`?
Can we reuse this model?

File: Adyen/Assets/zh-CN.lproj/Localizable.strings [hunk 32]
```diff
@@ -73,8 +72,6 @@
 "adyen.button.copy" = "复制";
 "adyen.voucher.introduction" = "感谢您的购买，请使用以下信息完成支付。";
 "adyen.voucher.readInstructions" = "阅读说明";
-"adyen.voucher.saveImage" = "";
-"adyen.voucher.finish" = "";
```
Comment (descorp) [inline Adyen/Assets/zh-CN.lproj/Localizable.strings]: This seems odd.. Maybe revert those lines for now - we can figure out what is wrong with them in next PR

File: AdyenComponents/Issuer List/IssuerListComponent.swift [hunk 33]
```diff
@@ -53,23 +52,25 @@ public final class IssuerListComponent: PaymentComponent, PaymentAware, Presenta
...
-                                                        paymentMethod: issuerListPaymentMethod,
-                                                        environment: context.apiContext.environment)
+                                                           localizedParameters: configuration.localizationParameters,
+                                                           paymentMethod: issuerListPaymentMethod,
+                                                           environment: context.apiContext.environment)
+
             listItem.selectionHandler = { [weak self] in
                 guard let self = self else { return }
                 
+                // open a intermediate view then on click of continue button open issuers details list
```
Comment (descorp) [inline AdyenComponents/Issuer List/IssuerListComponent.swift]: Is this comment still relevant?

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift [hunk 34]
```diff
@@ -0,0 +1,142 @@
...
+public final class OnlineBankingComponent: PaymentComponent,
+                                           PresentableComponent,
+                                           LoadingComponent,
+                                           PaymentAware {
+
+    private enum ViewIdentifier {
+        static let termsAndConditionsLabelItem = "OnlineBankingTermsAndConditionLabel"
+        static let continueButtonItem = "continueButton"
+        static let issuerListItem = "issuersList"
+    }
```
Comment (descorp) [inline AdyenComponents/OnlineBanking/OnlineBankingComponent.swift:20]: This is nice :)

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift [hunk 35]
```diff
@@ -0,0 +1,142 @@
...
+    public var configuration: Configuration
+
+    private let onlineBankingPaymentMethod: OnlineBankingPaymentMethod
+
+    private var selectedIssuer: Issuer
+
+    // MARK: - Items
+
+    private var termsAndConditionsLink: String {
+        return paymentMethod.type == .onlineBankingCZ ? "https://static.payu.com/sites/terms/files/payu_privacy_policy_cs.pdf" : "https://static.payu.com/sites/terms/files/payu_privacy_policy_sk.pdf"
```
Comment (descorp) [inline AdyenComponents/OnlineBanking/OnlineBankingComponent.swift]: Abstract question:

@marcperez
Do we have to keep those links in SDK?
Can backend provide it to us? 
Can it change in the future?

Comment (descorp) [inline AdyenComponents/OnlineBanking/OnlineBankingComponent.swift]: Maybe move this URLs it to an `enum`, like we do with `ViewIdentifier`?

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift [hunk 36]
```diff
@@ -0,0 +1,142 @@
...
+            self?.didSelectContinueButton()
+        }
+        return item
+    }()
+
+    /// The Issuer List item.
+    internal lazy var issuerListItem: FormIssuersPickerItem = {
+        let issuerListPickerItem: [IssuerPickerItem] = onlineBankingPaymentMethod.issuers.map({
+            return IssuerPickerItem(identifier: $0.identifier, element: $0)
+        })
```
Comment (descorp) [inline AdyenComponents/OnlineBanking/OnlineBankingComponent.swift]:
```suggestion
        let issuerListPickerItems = onlineBankingPaymentMethod.issuers.map { IssuerPickerItem(identifier: $0.identifier, element: $0)  }
```

File: Tests/Adyen Tests/Core/PaymentMethodTests.swift [hunk 37]
```diff
@@ -837,13 +837,3 @@ class PaymentMethodTests: XCTestCase {
         XCTAssertEqual(paymentMethod.name, "ACH Direct Debit")
     }
 }
-
-internal extension Coder {
```
Comment (descorp) [inline Tests/Adyen Tests/Core/PaymentMethodTests.swift:841]: Why is this deleted?

File: Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift [hunk 38]
```diff
@@ -0,0 +1,98 @@
...
+        let linkRanges = linkRangesInString()
+
+        let attributes = createAttributes(from: linkTextStyle)
+        for linkRange in linkRanges {
+            attributedString.addAttributes(attributes, range: linkRange)
+        }
+
+        label.attributedText = attributedString
+        label.accessibilityIdentifier = identifier
+        label.adyen.round(using: style.cornerRounding)
```
Comment (descorp) [inline Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift]: It is safer to apply `adyen.round(using: )` inside `layoutSubviews()` - corner radius can be sensitive to a view's height/width

File: Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift [hunk 39]
```diff
@@ -0,0 +1,98 @@
...
+    /// The style of the link text.
+    public var linkTextStyle: TextStyle
+
+    /// The text of the label.
+    public var originalText: String
+
+    // The link label.
+    public var link: String
+
+    public func build(with builder: FormItemViewBuilder) -> AnyFormItemView {
```
Comment (descorp) [inline Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift:42]: Minor: 
This could be useful in future!
Would be nice inject this logic into standalone "Utility" struct and cover with few more tests.

File: Tests/Adyen Tests/Core/PaymentMethodTests.swift (hunk 37 above)
Comment (neelSharma12) [reply Tests/Adyen Tests/Core/PaymentMethodTests.swift:841]: I moved to the separate extension file as this was needed in UITests as well.

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift (hunk 34 above)
Comment (neelSharma12) [reply AdyenComponents/OnlineBanking/OnlineBankingComponent.swift:20]: thanks :)

File: Adyen/Core/Payment Methods/Abstract/AnyPaymentMethodDecoder.swift (hunk 30 above)
Comment (neelSharma12) [reply Adyen/Core/Payment Methods/Abstract/AnyPaymentMethodDecoder.swift:281]: I have added this in "AnyPaymentMethodDecoder.swift". Please check :)

File: Adyen/Core/Payment Methods/OnlineBankingPaymentMethod.swift (hunk 31 above)
Comment (neelSharma12) [reply Adyen/Core/Payment Methods/OnlineBankingPaymentMethod.swift:10]: yes, we can :).

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift (hunk 27 above)
Comment (erenbesel) [reply AdyenComponents/OnlineBanking/OnlineBankingComponent.swift]: Shouldn't it be `issuerListPickerItems` ?

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift [hunk 40]
```diff
@@ -0,0 +1,142 @@
...
+    // MARK: - Items
+
+    private var termsAndConditionsLink: String {
+        return paymentMethod.type == .onlineBankingCZ ? "https://static.payu.com/sites/terms/files/payu_privacy_policy_cs.pdf" : "https://static.payu.com/sites/terms/files/payu_privacy_policy_sk.pdf"
+    }
+
+    /// The terms and condition message item.
+    internal lazy var termsAndConditionsLabelItem: FormAttributedLabelItem = .init(originalText:
+                        localizedString(.onlineBankingTermsAndConditions,
+                        configuration.localizationParameters),
```
Comment (erenbesel) [inline AdyenComponents/OnlineBanking/OnlineBankingComponent.swift]: I think there is indentation issue here

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift (hunk 35 above)
Comment (neelSharma12) [reply AdyenComponents/OnlineBanking/OnlineBankingComponent.swift]: Good suggestion :)

File: Adyen/UI/Form/Items/Value Pickers/Issuer List Picker/FormIssuerPickerItem.swift (hunk 12 above)
Comment (mohammedDehairy) [reply Adyen/UI/Form/Items/Value Pickers/Issuer List Picker/FormIssuerPickerItem.swift:13]: is it abstract? I don't think its abstract

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift [hunk 41]
```diff
@@ -0,0 +1,142 @@
...
+        AdyenAssertion.assert(message: "Issuer list should not be empty", condition: issuerListPickerItem.count <= 0)
+
+        let issuerPickerItem = FormIssuersPickerItem(preselectedValue: issuerListPickerItem[0],
+                                        selectableValues: issuerListPickerItem,
+                                        style: configuration.style.textField)
+        issuerPickerItem.title = localizedString(.selectFieldTitle, configuration.localizationParameters)
+        issuerPickerItem.identifier = ViewIdentifierBuilder.build(scopeInstance: self,
+                                                                  postfix: ViewIdentifier.issuerListItem)
+        observe(issuerPickerItem.publisher) { issuerPickerITem in
+            self.selectedIssuer = issuerPickerITem.element
```
Comment (erenbesel) [inline AdyenComponents/OnlineBanking/OnlineBankingComponent.swift]: Weakify self please :)

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift [hunk 42]
```diff
@@ -0,0 +1,142 @@
...
+        }
```
Comment (mohammedDehairy) [inline AdyenComponents/OnlineBanking/OnlineBankingComponent.swift]: I don't think you need to observe it, just get the `issuerPickerItem.value` when submitting, right?

File: AdyenComponents/OnlineBanking/OnlineBankingComponent.swift [hunk 43]
```diff
@@ -0,0 +1,142 @@
...
+    /// - Parameter paymentMethod: The Online Banking payment method.
+    /// - Parameter context: The context object for this component.
+    /// - Parameter configuration: The configuration for the component.
+    public init(paymentMethod: OnlineBankingPaymentMethod,
+                context: AdyenContext,
+                configuration: Configuration = .init()) {
+        self.onlineBankingPaymentMethod = paymentMethod
+        self.context = context
+        self.configuration = configuration
+        self.selectedIssuer = onlineBankingPaymentMethod.issuers[0]
```
Comment (erenbesel) [inline AdyenComponents/OnlineBanking/OnlineBankingComponent.swift]: bounds check.
Would be also nice to confirm from backend if this array can never be empty. If it can, I would go with guard break and optional `selectedIssuer` etc.
If it cannot, then assertion would be fine

File: Adyen/UI/Form/Items/Value Pickers/Issuer List Picker/FormIssuerPickerItem.swift (hunk 12 above)
Comment (neelSharma12) [reply Adyen/UI/Form/Items/Value Pickers/Issuer List Picker/FormIssuerPickerItem.swift:13]: It is a open class that needs to be subclassed. We are subclassing it for **FormPhoneExtensionPickerItem**  and **FormRegionPickerItem**

**Review by mohammedDehairy — DISMISSED** [review]

File: Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift (hunk 39 above)
Comment (neelSharma12) [reply Adyen/UI/Form/Items/Basic Items/FormAttributedLabelItem.swift:42]: As discussed, we will add tests for different strings in new PR.

**Review by erenbesel — APPROVED** [review]

**Review by mohammedDehairy — APPROVED** [review]

**Review by descorp — APPROVED** [review]
